server/common.ts

```diff
@@ -406,14 +418,49 @@ export function authenticateWithReject(request: express.Request, response: expre
 	}
 }
 // For directly user facing endpoints
-export function authenticateWithRedirect(request: express.Request, response: express.Response, next: express.NextFunction) {
+export async function authenticateWithRedirect(request: express.Request, response: express.Response, next: express.NextFunction) {
 	if (!request.isAuthenticated()) {
-		response.redirect("/login");
+		await redirectToLogin(request, response);
 	}
 	else {
 		next();
 	}
 }
+
+export function getExternalPort(request: express.Request): number {
+	function defaultPort(): number {
+		// Default ports for HTTP and HTTPS
+		return request.protocol === "http" ? 80 : 443;
+	}
+
+	let host = request.headers.host;
+	if (!host || Array.isArray(host)) {
+		return defaultPort();
+	}
+
+	// IPv6 literal support
+	let offset = host[0] === "[" ? host.indexOf("]") + 1 : 0;
+	let index = host.indexOf(":", offset);
+	if (index !== -1) {
+		return parseInt(host.substring(index + 1), 10);
+	}
+	else {
+		return defaultPort();
+	}
+}
+
+export function createLink(request: express.Request, link: string): string {
```

Comment:
Since this is being called in more and different places than previously, we need some way of ensuring the hostname is valid.